### PR TITLE
Change password

### LINKED ISSUE #279: Implement endpoint and handler for change password

### PR DESCRIPTION
**Pull request by Forestieur**
## Issue ticket number and link
closes #279
## Checklist before requesting a review
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Have you lint your code locally before submission?
- [x] If that PR have tests
- [x] I have performed a self-review of my code

### CODE REVIEW

File: Streetcode/Streetcode.BLL/MediatR/Account/ChangePassword/ChangePasswordHandler.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            var result = await _userManager.ChangePasswordAsync(user, request.PasswordChange.CurrentPassword, request.PasswordChange.NewPassword);
+
+            if (!result.Succeeded)
+            {
+                var errorMsg = MessageResourceContext.GetMessage(ErrorMessages.FailToChangePassword, request);
+                _logger.LogError(request, errorMsg);
+                return Result.Fail(new Error(errorMsg));
+            }
+
+            // remove tokens cookies and logout after successfully changing password
```
Comment (metrondir) [inline Streetcode/Streetcode.BLL/MediatR/Account/ChangePassword/ChangePasswordHandler.cs:72]: erase that line

File: Streetcode/Streetcode.BLL/MediatR/Account/ChangePassword/ChangePasswordHandler.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+            var cacheResult = await _cacheService.SetBlacklistedTokenAsync(accessToken, user.Id.ToString());
+            if (!cacheResult)
+            {
+                var errorMsg = MessageResourceContext.GetMessage(ErrorMessages.FailedToSetTokenInBlackList, request);
+                _logger.LogError(cacheResult, errorMsg);
+                return Result.Fail(new Error(errorMsg));
```
Comment (metrondir) [inline Streetcode/Streetcode.BLL/MediatR/Account/ChangePassword/ChangePasswordHandler.cs:79]: i think we should logout all devices with old password (it mean delete all refreshTokens from db that user)

**Review by metrondir — APPROVED** [review]
lgtm!

File: Streetcode/Streetcode.BLL/MediatR/Account/ChangePassword/ChangePasswordDTOValidator.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+namespace Streetcode.BLL.MediatR.Account.ChangePassword
+{
+    public class ChangePasswordDTOValidator : AbstractValidator<ChangePasswordCommand>
+    {
+        public ChangePasswordDTOValidator()
+        {
+            RuleFor(u => u.PasswordChange.NewPassword).MinimumLength(7);
+            RuleFor(u => u.PasswordChange).Must(x => x.NewPassword == x.ConfirmPassword);
+            RuleFor(u => u.PasswordChange).Must(x => x.NewPassword != x.CurrentPassword);
+        }
```
Comment (PavloIeremeiew) [inline Streetcode/Streetcode.BLL/MediatR/Account/ChangePassword/ChangePasswordDTOValidator.cs:12]: the password must contain an uppercase and lowercase letter, a number and a special character

File: Streetcode/Streetcode.BLL/MediatR/Account/ChangePassword/ChangePasswordHandler.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        private readonly UserManager<User> _userManager;
+        private readonly ICacheService _cacheService;
+        private readonly ITokenService _tokenService;
+        private readonly ILoggerService _logger;
+        private readonly IHttpContextAccessor _httpContextAccessor;
+        private readonly ICookieService _cookieService;
+
+        public ChangePasswordHandler(
+            UserManager<User> userManager,
+            ICacheService cacheService,
```
Comment (serhiiyasenev) [inline Streetcode/Streetcode.BLL/MediatR/Account/ChangePassword/ChangePasswordHandler.cs:26]: it seems we don't need `ICacheService cacheService` here

Comment (Forestieur) [reply Streetcode/Streetcode.BLL/MediatR/Account/ChangePassword/ChangePasswordHandler.cs:26]: cache service is used to SetBlacklistedTokenAsync on user access token

File: Streetcode/Streetcode.BLL/MediatR/Account/ChangePassword/ChangePasswordDTOValidator.cs (hunk 3 above)
Comment (Forestieur) [reply Streetcode/Streetcode.BLL/MediatR/Account/ChangePassword/ChangePasswordDTOValidator.cs:18]: added rules

**Review by PavloIeremeiew — APPROVED** [review]

**Pull request merged**